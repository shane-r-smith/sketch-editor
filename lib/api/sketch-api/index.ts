export * from "./sketch-api";
export * from "./sketch-api.types";
