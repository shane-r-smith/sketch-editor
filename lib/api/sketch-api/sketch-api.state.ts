import { atom } from "jotai";
import {
  type Layer,
  type Page,
  type Sketch,
  type SketchElement,
} from "../../domain";

export const sketchAtom = atom<Sketch>();
export const currentPageIndexAtom = atom(0);
export const currentLayerIndexAtom = atom(0);

export const currentPageAtom = atom(
  (get) => {
    const sketch = get(sketchAtom);
    const pageIndex = get(currentPageIndexAtom);

    return sketch?.pages[pageIndex];
  },
  (get, set, newPage: Partial<Page>) => {
    const sketch = get(sketchAtom);

    if (!sketch) {
      // no sketch to update
      return;
    }

    const updated = structuredClone(sketch);

    updated.pages = sketch.pages.map<Page>((page) => {
      if (page.id === newPage.id) {
        return { ...page, ...newPage };
      }

      return page;
    });

    set(sketchAtom, updated);
  }
);

export const currentLayersAtom = atom((get) => {
  const page = get(currentPageAtom);
  return page?.layers;
});

export const currentLayerAtom = atom(
  (get) => {
    const page = get(currentPageAtom);
    const layerIndex = get(currentLayerIndexAtom);

    return page?.layers[layerIndex];
  },
  (get, set, newLayer: Partial<Layer>) => {
    const page = get(currentPageAtom);

    if (!page || !newLayer.id) {
      // no page to update
      return;
    }

    page.layers = page.layers.map<Layer>((layer: Layer) => {
      if (layer.id === newLayer.id && newLayer.id !== undefined) {
        return { ...layer, ...newLayer } as Layer;
      }

      return layer;
    });

    set(currentPageAtom, page);
  }
);

export const addElementAtom = atom(
  null,
  (get, set, newElement: SketchElement) => {
    const currentLayer = get(currentLayerAtom);

    if (!currentLayer) {
      // no current layer to update
      return;
    }

    currentLayer.elements = [
      ...currentLayer.elements,
      normaliseElement(newElement),
    ];

    set(currentLayerAtom, currentLayer);
  }
);

function normaliseElement(element: SketchElement): SketchElement {
  if (element.tool === "PEN") {
    delete (element as unknown as { strokeColourHistory: unknown })
      .strokeColourHistory;
  }

  delete (element as unknown as { icon: unknown }).icon;

  return element;
}
