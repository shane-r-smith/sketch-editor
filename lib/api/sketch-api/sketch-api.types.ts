import type { Sketch } from "../../domain";
import type { IEventsApi } from "../events-api";

export interface ISketchApi extends IEventsApi {
  load: (sketch: Sketch) => void;
  create: (name: string, size: [number, number], frameRate: number) => Sketch;
  clear: () => void;
}
