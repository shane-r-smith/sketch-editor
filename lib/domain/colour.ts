export type ColourType = "HEX" | "RGB";
