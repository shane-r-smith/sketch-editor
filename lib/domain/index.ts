export * from "./colour";
export * from "./drawing";
export * from "./sketch-element";
export * from "./tool";
export * from "./layer";
export * from "./page";
export * from "./sketch";
