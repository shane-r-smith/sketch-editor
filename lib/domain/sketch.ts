import { Page } from "./page";
import { v6 as uuidv6 } from "uuid";

export interface SketchMetadata {
  id: string;
  name: string;
  size: [number, number];
  created: string;
  edited: string;
  frameRate: number;
}

export class Sketch implements SketchMetadata {
  id: string = uuidv6();
  created: string = new Date().toISOString();
  edited: string = new Date().toISOString();
  pages: Page[] = [new Page()];

  name: string;
  size: [number, number];
  frameRate: number;

  constructor(name: string, size: [number, number], frameRate: number) {
    this.name = name;
    this.size = size;
    this.frameRate = frameRate;
  }
}
