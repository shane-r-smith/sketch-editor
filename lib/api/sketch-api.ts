import { Sketch } from "../domain";
import type { ISketchApi } from "./sketch-api.types";
import { sketchAtom } from "./sketch.state";
import { sketchStore } from "../sketch.store";

export class SketchApi implements ISketchApi {
  public load(sketch: Sketch): void {
    sketchStore.set(sketchAtom, sketch);
  }

  public create(name: string, size: [number, number], frameRate: number) {
    return new Sketch(name, size, frameRate);
  }

  public clear() {
    sketchStore.set(sketchAtom, undefined);
  }
}
