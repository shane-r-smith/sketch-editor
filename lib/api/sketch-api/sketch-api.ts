import { Sketch } from "../../domain";
import type { ISketchApi } from "./sketch-api.types";
import { sketchAtom } from "./sketch-api.state";
import { sketchStore } from "../../sketch.store";
import { EventsApi } from "../events-api";

export class SketchApi extends EventsApi implements ISketchApi {
  public load(sketch: Sketch): void {
    sketchStore.set(sketchAtom, sketch);
  }

  public create(name: string, size: [number, number], frameRate: number) {
    return new Sketch(name, size, frameRate);
  }

  public clear() {
    sketchStore.set(sketchAtom, undefined);
  }
}
