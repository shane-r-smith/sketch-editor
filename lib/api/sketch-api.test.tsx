import { Sketch } from "../domain";
import { SketchApi } from "./sketch-api";
import { sketchStore } from "../sketch.store";
import { sketchAtom } from "./sketch.state";

vi.mock("uuid", () => ({
  v6: () => "mocked-unique-id",
}));

describe("SketchApi", () => {
  let api: SketchApi;

  beforeEach(() => {
    api = new SketchApi();
  });

  describe("load", () => {
    it("sets the sketch in the store", () => {
      const sketch = new Sketch("Test", [100, 100], 60);
      api.load(sketch);

      expect(sketchStore.get(sketchAtom)).toMatchSnapshot();
    });
  });

  describe("create", () => {
    it("creates a new Sketch instance with given parameters", () => {
      const name = "MySketch";
      const size: [number, number] = [200, 150];
      const frameRate = 30;
      const sketch = api.create(name, size, frameRate);

      expect(sketch).toMatchSnapshot();
    });
  });

  describe("clear", () => {
    it("sets the sketch in the store to undefined", () => {
      const sketch = new Sketch("Test", [100, 100], 60);
      api.load(sketch);

      expect(sketchStore.get(sketchAtom)).not.toBeUndefined();

      api.clear();

      expect(sketchStore.get(sketchAtom)).toBeUndefined();
    });
  });
});
