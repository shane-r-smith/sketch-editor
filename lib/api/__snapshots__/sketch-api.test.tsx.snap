// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SketchApi > create > creates a new Sketch instance with given parameters 1`] = `
Sketch {
  "created": "2025-08-24T14:25:52.820Z",
  "edited": "2025-08-24T14:25:52.820Z",
  "frameRate": 30,
  "id": "mocked-unique-id",
  "name": "MySketch",
  "pages": [
    Page {
      "id": "mocked-unique-id",
      "layers": [
        Layer {
          "elements": [],
          "id": "mocked-unique-id",
          "locked": false,
          "name": "Layer 1",
          "opacity": 1,
          "visible": true,
        },
      ],
    },
  ],
  "size": [
    200,
    150,
  ],
}
`;

exports[`SketchApi > load > sets the sketch in the store 1`] = `
Sketch {
  "created": "2025-08-24T14:25:52.818Z",
  "edited": "2025-08-24T14:25:52.818Z",
  "frameRate": 60,
  "id": "mocked-unique-id",
  "name": "Test",
  "pages": [
    Page {
      "id": "mocked-unique-id",
      "layers": [
        Layer {
          "elements": [],
          "id": "mocked-unique-id",
          "locked": false,
          "name": "Layer 1",
          "opacity": 1,
          "visible": true,
        },
      ],
    },
  ],
  "size": [
    100,
    100,
  ],
}
`;
